### PR TITLE
Migrate `#[allow(...)]` to `#[expect(...)]`

### LINKED ISSUE #15059: Swap most uses of allow in lints to expect
## What problem does this solve or what need does it fill?

With the advent of [Rust 1.81](https://blog.rust-lang.org/2024/09/05/Rust-1.81.0.html), we can now distinguish between lints that we haven't gotten around to fixing and those where the lint is wrong.

## What solution would you like?

Most localized lints in Bevy should be `expect`: the only exception that comes to mind is `missing_docs`.

Leave the `Cargo.toml` lints set as allow, as funny as it would be to set a crate-level `expect(type_complexity)`.

### PR DESCRIPTION
**Pull request by BD103**
# Objective

- Rust 1.81 released the `#[expect(...)]` attribute, which works like `#[allow(...)]` but throws a warning if the lint _isn't_ raised. This is preferred to `#[allow(...)]` because it tells us when it can be removed.
- Closes #15059.

## Solution

Switch all applicable `#[allow(...)]` attributes to `#[expect(...)]` with a reason attribute.

### Progress

- [x] `bevy_ptr`
- [x] `bevy_utils`
- [x] `bevy_reflect`
- [x] `bevy_tasks`
- [x] `bevy_ecs`
- [x] `bevy_derive`
- [x] `bevy_app`
- [x] `bevy_reflect_derive`
- [x] `bevy_asset_macros`
- [x] `bevy_encase_derive`
- [x] `bevy_render_macros`
- [x] `bevy_state_macros`
- [x] `bevy_log`
- [x] `bevy_time`
- [x] `bevy_diagnostic`
- [x] `bevy_transform`
- [x] `bevy_asset`
- [ ] `bevy_input`
- [ ] `bevy_gilrs`
- [ ] `bevy_window`
- [ ] `bevy_winit`
- [ ] `bevy_state`
- [ ] `bevy_render`
- [ ] `bevy_picking`
- [ ] `bevy_core_pipeline`
- [ ] `bevy_sprite`
- [ ] `bevy_text`
- [ ] `bevy_pbr`
- [ ] `bevy_ui`
- [ ] `bevy_gltf`
- [ ] `bevy_gizmos`
- [ ] `bevy_dev_tools`
- [ ] `bevy_internal`
- [ ] `bevy_dylib`

## Testing

- `cargo clippy --workspace --all-features`

### CODE REVIEW

File: crates/bevy_reflect/derive/src/derive_data.rs [hunk 1]
```diff
@@ -121,10 +121,9 @@ pub(crate) struct EnumVariant<'a> {
     /// The fields within this variant.
     pub fields: EnumVariantFields<'a>,
     /// The reflection-based attributes on the variant.
-    #[allow(dead_code)]
     pub attrs: FieldAttributes,
     /// The index of this variant within the enum.
-    #[allow(dead_code)]
+    #[expect(dead_code, reason = "TODO(MrGVSV): Why should this be kept?")]
```
Comment (MrGVSV) [inline crates/bevy_reflect/derive/src/derive_data.rs:126]: It probably doesn't need to be kept anymore 😅

File: crates/bevy_reflect/derive/src/reflect_value.rs [hunk 2]
```diff
@@ -25,7 +25,7 @@ use syn::{parenthesized, Attribute, Generics, Path};
 /// (in my_crate::bar) Bar(TraitA, TraitB)
 /// ```
 pub(crate) struct ReflectValueDef {
-    #[allow(dead_code)]
+    #[expect(dead_code, reason = "TODO(MrGVSV): Why should this be kept?")]
```
Comment (MrGVSV) [inline crates/bevy_reflect/derive/src/reflect_value.rs:28]: I think this field is also unneeded now.

File: crates/bevy_app/src/plugin.rs [hunk 3]
```diff
@@ -167,7 +167,9 @@ mod sealed {
             where
                 $($plugins: Plugins<$param>),*
             {
-                #[allow(non_snake_case, unused_variables)]
+                // We use `allow` instead of `expect` here because the lint is not generated for all cases.
+                #[allow(non_snake_case, reason = "`all_tuples!()` generates non-snake-case variable names.")]
+                #[allow(unused_variables, reason = "`app` is unused when implemented for unit `()` type.")]
```
Comment (benfrankel) [inline crates/bevy_app/src/plugin.rs:172]:
```suggestion
                #[allow(unused_variables, reason = "`app` is unused when implemented for the unit type `()`.")]
```
Optional nit

File: crates/bevy_reflect/derive/src/reflect_value.rs (hunk 2 above)
Comment (BD103) [reply crates/bevy_reflect/derive/src/reflect_value.rs]: I just found out that this field is used when the `documentation` feature is enabled, so I'm going to keep it.

Comment (MrGVSV) [reply crates/bevy_reflect/derive/src/reflect_value.rs]: Oh good to know! This is exactly why this PR is going to be really useful haha

**Comment by BD103** [comment]
I'm going to put this PR up for adoption. I no longer have the energy to work on it, and there are still a _lot_ of crates left. If you want to adopt this, my workflow looked like this:

1. Run `cargo clippy --workspace --message-format=short`
2. `cd` into the first crate that raises warnings.
3. Run `cargo watch -c -x clippy` until warnings stop appearing.
4. Run `cargo clippy --no-default-features` and `cargo clippy --all-features` for good measure.
5. Repeat step 1 until there are no more warning.

**Comment by richchurcher** [comment]
@BD103 I don't know if I can do the rest of this justice, but I figured I'd at least try to clean up the conflicts, it'd be a crime to see all that go to waste :laughing: #15301

**Pull request closed without merging**